iocell/src/main/scala/chisel/IOCell.scala

```diff
@@ -0,0 +1,188 @@
+    anaFn: () => AnalogIOCell = IOCell.exampleAnalog): (T, Seq[IOCell]) =
+  {
+    val padSignal = IO(DataMirror.internal.chiselTypeClone[T](coreSignal))
+    val iocells = IOCell.generateFromSignal(coreSignal, padSignal, name, inFn, outFn, anaFn)
+    (padSignal, iocells)
+  }
+
+  def generateFromSignal[T <: Data](coreSignal: T, padSignal: T, name: Option[String] = None,
+    inFn: () => DigitalInIOCell = IOCell.exampleInput,
+    outFn: () => DigitalOutIOCell = IOCell.exampleOutput,
+    anaFn: () => AnalogIOCell = IOCell.exampleAnalog): Seq[IOCell] =
+  {
+    coreSignal match {
+      case coreSignal: Analog => {
+        if (coreSignal.getWidth == 0) {
+          Seq()
+        } else {
+          require(coreSignal.getWidth == 1, "Analogs wider than 1 bit are not supported because we can't bit-select Analogs (https://github.com/freechipsproject/chisel3/issues/536)")
+          val iocell = anaFn()
+          name.foreach(n => iocell.suggestName(n))
+          iocell.io.core <> coreSignal
+          padSignal <> iocell.io.pad
+          Seq(iocell)
+        }
+      }
+      case coreSignal: Clock => {
+        DataMirror.directionOf(coreSignal) match {
+          case ActualDirection.Input => {
+            val iocell = inFn()
+            name.foreach(n => iocell.suggestName(n))
+            coreSignal := iocell.io.i.asClock
+            iocell.io.ie := true.B
+            iocell.io.pad := padSignal.asUInt.asBool
+            Seq(iocell)
+          }
+          case ActualDirection.Output => {
+            val iocell = outFn()
+            name.foreach(n => iocell.suggestName(n))
+            iocell.io.o := coreSignal.asUInt.asBool
+            iocell.io.oe := true.B
+            padSignal := iocell.io.pad.asClock
+            Seq(iocell)
+          }
+          case _ => throw new Exception("Unknown direction")
+        }
+      }
+      case coreSignal: Bits => {
+        require(padSignal.getWidth == coreSignal.getWidth, "padSignal and coreSignal must be the same width")
+        if (padSignal.getWidth == 0) {
+          // This dummy assignment will prevent invalid firrtl from being emitted
+          DataMirror.directionOf(coreSignal) match {
+            case ActualDirection.Input => coreSignal := 0.U
+            case _ => {}
+          }
+          Seq()
+        } else {
+          DataMirror.directionOf(coreSignal) match {
+            case ActualDirection.Input => {
+              // this type cast is safe because we guarantee that padSignal and coreSignal are the same type (T), but the compiler is not smart enough to know that
+              val iocells = padSignal.asInstanceOf[Bits].asBools.zipWithIndex.map { case (w, i) =>
+                val iocell = inFn()
+                name.foreach(n => iocell.suggestName(n + "_" + i))
+                iocell.io.pad := w
+                iocell.io.ie := true.B
+                iocell
+              }
+              coreSignal := Cat(iocells.map(_.io.i).reverse)
```

Comment:
Should definitely have a comment about the `reverse` here, maybe also point to where the similar thing is done in chisel/firrtl.

Reply:
OK, I'll add a comment. This is mostly just a bad Chisel API, IMO, but I understand "Why"

Reply:
Done